internal/worker/logsink/loggers.go

```diff
@@ -4,110 +4,214 @@
 package logsink
 
 import (
+	"context"
 	"io"
-	"strings"
+	"path/filepath"
+	"strconv"
 	"sync"
 	"time"
 
 	"github.com/juju/clock"
 	"github.com/juju/errors"
+	"github.com/juju/loggo/v2"
+	"github.com/juju/worker/v4"
+	"gopkg.in/tomb.v2"
 
+	"github.com/juju/juju/core/logger"
 	corelogger "github.com/juju/juju/core/logger"
+	internallogger "github.com/juju/juju/internal/logger"
 )
 
-type bufferedLogWriterCloser struct {
-	*corelogger.BufferedLogWriter
-	closer io.Closer
+var (
+	fallbackLogger = internallogger.GetLogger("logsink")
+)
+
+type modelLogger struct {
+	tomb tomb.Tomb
+
+	bufferedLogWriter *bufferedLogWriterCloser
+	bufferedLogCloser func() error
+
+	loggerContext corelogger.LoggerContext
 }
 
-func (b *bufferedLogWriterCloser) Close() error {
-	err := errors.Trace(b.BufferedLogWriter.Flush())
-	_ = b.closer.Close()
-	return err
+// ModelLoggerConfig holds the configuration for a model logger.
+type ModelLoggerConfig struct {
+	MachineID     string
+	NewLogWriter  corelogger.LogWriterForModelFunc
+	BufferSize    int
+	FlushInterval time.Duration
+	Clock         clock.Clock
 }
 
 // NewModelLogger returns a new model logger instance.
 // The actual loggers returned for each model are created
 // by the supplied loggerForModelFunc.
 func NewModelLogger(
-	loggerForModelFunc corelogger.LogWriterForModelFunc,
-	bufferSize int,
-	flushInterval time.Duration,
-	clock clock.Clock,
-) corelogger.ModelLogger {
-	return &modelLogger{
-		clock:               clock,
-		loggerBufferSize:    bufferSize,
-		loggerFlushInterval: flushInterval,
-		loggerForModel:      loggerForModelFunc,
-		modelLoggers:        make(map[string]corelogger.LogWriterCloser),
-	}
+	ctx context.Context,
+	key corelogger.LoggerKey,
+	config ModelLoggerConfig,
+) (worker.Worker, error) {
+	return newModelLogger(ctx, key, config)
 }
 
-type modelLogger struct {
-	mu sync.Mutex
-
-	clock               clock.Clock
-	loggerBufferSize    int
-	loggerFlushInterval time.Duration
-
-	modelLoggers   map[string]corelogger.LogWriterCloser
-	loggerForModel corelogger.LogWriterForModelFunc
-}
-
-// GetLogWriter creates a new log writer for the given model UUID.
-func (d *modelLogger) GetLogWriter(modelUUID, modelName, modelOwner string) (corelogger.LogWriterCloser, error) {
-	d.mu.Lock()
-	defer d.mu.Unlock()
-	if l, ok := d.modelLoggers[modelUUID]; ok {
-		return l, nil
-	}
```

Comment:
This issue of having a modelUUID and a different modelName and modelOwner is a thing of the past now. We take that into account.